pyrokinetics/local_geometry/LocalGeometryFourier.py

```diff
@@ -0,0 +1,662 @@
+        drho_dpsi = eq.rho.derivative()(psi_n)
+        shift = eq.R_major.derivative()(psi_n) / drho_dpsi / eq.a_minor
+        dZ0dr = eq.Z_mid.derivative()(psi_n) / drho_dpsi / eq.a_minor
+
+        pressure = eq.pressure(psi_n)
+        q = eq.q(psi_n)
+
+        dp_dpsi = eq.q.derivative()(psi_n)
+
+        shat = rho / q * dp_dpsi / drho_dpsi
+
+        dpressure_drho = eq.p_prime(psi_n) / drho_dpsi
+
+        beta_prime = 8 * pi * 1e-7 * dpressure_drho / B0**2
+
+        normalised_height = (Z - Zmid) / (kappa * r_minor)
+
+        # Floating point error can lead to >|1.0|
+        normalised_height = np.where(
+            np.isclose(normalised_height, 1.0), 1.0, normalised_height
+        )
+        normalised_height = np.where(
+            np.isclose(normalised_height, -1.0), -1.0, normalised_height
+        )
+
+        theta = np.arcsin(normalised_height)
+
+        normalised_radius = (R - R_major) / r_minor
+
+        normalised_radius = np.where(
+            np.isclose(normalised_radius, 1.0), 1.0, normalised_radius
+        )
+        normalised_radius = np.where(
+            np.isclose(normalised_radius, -1.0), -1.0, normalised_radius
+        )
+
+        thetaR = np.arccos(normalised_radius)
+
+        for i in range(len(theta)):
+            if R[i] < R_upper:
+                theta[i] = np.pi - theta[i]
+            else:
+                if Z[i] < 0:
+                    theta[i] = 2 * np.pi + theta[i]
+
+        for i in range(len(thetaR)):
+            if Z[i] < 0:
+                thetaR[i] = 2 * np.pi - thetaR[i]
+
+        theta_diff = thetaR - theta
+
+        self.psi_n = psi_n
+        self.rho = float(rho)
+        self.r_minor = float(r_minor)
+        self.Rmaj = float(R_major / eq.a_minor)
+        self.a_minor = float(eq.a_minor)
+        self.f_psi = float(fpsi)
+        self.B0 = float(B0)
+
+        self.Z0 = float(Zmid / eq.a_minor)
+        self.R = R
+        self.Z = Z
+        self.theta = theta
+        self.b_poloidal = b_poloidal
+
+        self.q = float(q)
+        self.shat = shat
+        self.beta_prime = beta_prime
+        self.pressure = pressure
+        self.dpressure_drho = dpressure_drho
+
+        for n_moments in [4, 5, 6, 8, 10, 12, 14, 16]:
+            # n_moments = 8
+
+            asym_coeff = np.empty(n_moments)
+            sym_coeff = np.empty(n_moments)
+
+            for i in range(n_moments):
+                c_int = theta_diff * np.cos(i * theta)
+                asym_coeff[i] = simpson(c_int, theta)
+
+                s_int = theta_diff * np.sin(i * theta)
+                sym_coeff[i] = simpson(s_int, theta)
+
+            asym_coeff *= 1 / np.pi
+            sym_coeff *= 1 / np.pi
+
+            self.kappa = kappa
+            self.n_moments = n_moments
+            self.sym_coeff = sym_coeff
+            self.asym_coeff = asym_coeff
+
+            self.thetaR = self.get_thetaR(theta)
+            self.dthetaR_dtheta = self.get_dthetaR_dtheta(theta)
+
+            """
```

Comment:
I'm guessing this is just here for debugging during development?

Reply:
Quite a few functions use them so might be safer to keep them as attrs but not necessary